### PR TITLE
Rya 101 and 102 Mongo Rya example fails to initialize.

### PR DESCRIPTION
**Pull request by DLotts**
Initial fix.  Blocked by RYA-133.  More commits to come.

### CODE REVIEW

File: dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java [hunk 1]
```diff
@@ -41,26 +47,77 @@
...
+                    MongodForTestsFactory testsFactory = MongodForTestsFactory.with(Version.Main.PRODUCTION);
+                    mongoClient = testsFactory.newMongo();
+                } catch (IOException e) {
+                    // Rethrow as an unchecked error.  Since we are in a test mode here, just fail fast.
+                    throw new MongoException("While creating a factory for a test/mock MongoDB instance.",e);
+                }
             } else {
-                mongoClient = new MongoClient(server);
+                // Connect to a running Mongo server
+                final String host = requireNonNull(conf.get(MongoDBRdfConfiguration.MONGO_INSTANCE), "MongoDB host name is required");
```
Comment (amihalik) [inline dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java]: Use guava preconditions instead (e.g. [here](https://github.com/apache/incubator-rya/blob/7743a42a5febe440107e037aadc3753643708e83/common/rya.api/src/main/java/mvm/rya/api/RdfCloudTripleStoreConfiguration.java#L167))

File: dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java [hunk 2]
```diff
@@ -41,26 +47,77 @@
...
-            if (conf.get(MongoDBRdfConfiguration.MONGO_USER) != null) {
-                final String username = conf.get(MongoDBRdfConfiguration.MONGO_USER);
-                final String dbName = conf.get(MongoDBRdfConfiguration.MONGO_DB_NAME);
-                final char[] pswd = conf.get(MongoDBRdfConfiguration.MONGO_USER_PASSWORD).toCharArray();
-                final MongoCredential cred = MongoCredential.createCredential(username, dbName, pswd);
-                mongoClient = new MongoClient(server, Arrays.asList(cred));
+    public static synchronized MongoClient getMongoClient(final Configuration conf)
+            throws ConfigurationRuntimeException, MongoException {
+        if (mongoClient == null) {
+            // The singleton client has not yet created, is it a test/mock instance, or a service?
```
Comment (isper3at) [inline dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java]: this doc is misleading.  There is only one, but it isn't a singleton.

File: dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java [hunk 3]
```diff
@@ -41,26 +47,77 @@
...
+            if (conf.getBoolean(MongoDBRdfConfiguration.USE_TEST_MONGO, false)) {
```
Comment (isper3at) [inline dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java]: break these out into their own methods, getTestMongoClient()/getMongoClient()

File: dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java [hunk 4]
```diff
@@ -41,26 +47,77 @@
...
+                } catch (UnknownHostException e) {
+                    throw new MongoException("Connecting MongoDB instance, cannot find host="+host,e);
+                }
+                // check for authentication credentials
+                if (conf.get(MongoDBRdfConfiguration.MONGO_USER) != null) {
+                    final String username = conf.get(MongoDBRdfConfiguration.MONGO_USER);
+                    final String dbName = requireNonNull(conf.get(MongoDBRdfConfiguration.MONGO_DB_NAME),
+                            MongoDBRdfConfiguration.MONGO_DB_NAME + " is null but required configuration if "
+                                    + MongoDBRdfConfiguration.MONGO_USER + " is configured.");
+                    final char[] pswd = requireNonNull(conf.get(MongoDBRdfConfiguration.MONGO_USER_PASSWORD),
```
Comment (isper3at) [inline dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java]: the requireNonNull() should be before this.  assignment should only happen after the precondition checks.

File: dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java [hunk 5]
```diff
@@ -41,26 +47,77 @@
...
+
+    /**
+     * Throw exception for un-configured required values.
+     * 
+     * @param required  String to check
+     * @param message  throw configuration exception with this description
+     * @return unaltered required string
+     * @throws ConfigurationRuntimeException  if required is null
+     */
+    static private String requireNonNull(String required, String message) throws ConfigurationRuntimeException {
```
Comment (isper3at) [inline dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java]: what Aaron said.  This is not needed

File: dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java [hunk 6]
```diff
@@ -41,26 +47,77 @@
...
+        if (required == null)
+            throw new ConfigurationRuntimeException(message);
+        return required;
+    }
+
+    /*
+     * Same as above, check that it is a integer and return the parsed integer.
+     */
+    static private int requireNonNullInt(String required, String message) throws ConfigurationRuntimeException {
```
Comment (isper3at) [inline dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java]: same as above, as a general note scope is the first keyword, then static.

File: dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoDBRyaDAO.java [hunk 7]
```diff
@@ -73,14 +73,9 @@
...
         this.conf = conf;
-        try {
-            mongoClient = MongoConnectorFactory.getMongoClient(conf);
-            conf.setMongoClient(mongoClient);
-            init();
-        } catch (NumberFormatException | UnknownHostException e) {
-            log.error("Unable to create a connection to mongo.", e);
-            throw e;
-        }
+        mongoClient = MongoConnectorFactory.getMongoClient(conf);
```
Comment (isper3at) [inline dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoDBRyaDAO.java]: why did you remove the log?

File: extras/indexing/src/main/java/mvm/rya/indexing/FilterFunctionOptimizer.java [hunk 8]
```diff
@@ -99,29 +99,19 @@ public void setConf(final Configuration conf) {
         this.conf = conf;
         //reset the init.
         init = false;
-        try {
             init();
-        } catch (final NumberFormatException | UnknownHostException e) {
-            LOG.error("Unable to update to use new config, falling back to the old config.", e);
```
Comment (isper3at) [inline extras/indexing/src/main/java/mvm/rya/indexing/FilterFunctionOptimizer.java]: same here

File: extras/indexing/src/main/java/mvm/rya/indexing/FilterFunctionOptimizer.java [hunk 9]
```diff
@@ -99,29 +99,19 @@ public void setConf(final Configuration conf) {
...
             if (ConfigUtils.getUseMongo(conf)) {
-                try {
                     final MongoClient mongoClient = MongoConnectorFactory.getMongoClient(conf);
                     geoIndexer = new MongoGeoIndexer(mongoClient);
                     geoIndexer.setConf(conf);
                     freeTextIndexer = new MongoFreeTextIndexer(mongoClient);
                     freeTextIndexer.setConf(conf);
                     temporalIndexer = new MongoTemporalIndexer(mongoClient);
                     temporalIndexer.setConf(conf);
-                } catch (NumberFormatException | UnknownHostException e) {
```
Comment (isper3at) [inline extras/indexing/src/main/java/mvm/rya/indexing/FilterFunctionOptimizer.java]: same here

File: dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java (hunk 1 above)
Comment (DLotts) [inline dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java]: I started with that.  (Objects.requireNonNull() is similar)   But I wanted to throw a different exception, not NullPointerException since the caller may want to specifically catch it.  So then I wrapped the whole thing in a try with a catch that re-throws.  But then it catches ANY null pointer, not just the ones I checked and it became more complicated, so I did it this way.  What do you think, is it worth it to throw a different unchecked error?

File: dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java (hunk 5 above)
Comment (DLotts) [inline dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java]: In addition to what I said to Aaron:  I am preserving the difference between a missing configuration and a random NullPointerException and NumberFormatException.  
I'll switch it back if you all consider this unimportant.  The method now throws two unchecked, and IMO not recoverable errors:  ConfigurationRuntimeException for configuration parameters that are missing, and MongoException for issues preventing connection, possibly temporary.  But those issues can be specifically caught, for example: to retry.   A NullPointerException could be thrown for any reason.

File: dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoDBRyaDAO.java (hunk 7 above)
Comment (DLotts) [inline dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoDBRyaDAO.java]: the UnknownHostException is a checked exception and must be caught.  I removed both of the exceptions and wrapped into better unchecked exceptions.  I assume that the default error handler will log the error.  Otherwise we should attempt to log every unchecked exception.  I made the underlying thrown errors more descriptive, they all mention that the MongoDB connection has failed, so this log entry is redundant.

File: dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java (hunk 1 above)
Comment (isper3at) [inline dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java]: This seems like an abuse of try/catch.  You generally do a check on the config values way before this point.  Doing it this way can also be super annoying if multiple values are missing.  Run, break, missing a.  Run, break, missing c.  Run, break, missing e, etc.

File: extras/indexing/src/main/java/mvm/rya/indexing/FilterFunctionOptimizer.java (hunk 8 above)
Comment (DLotts) [inline extras/indexing/src/main/java/mvm/rya/indexing/FilterFunctionOptimizer.java]: Same reason.  But in addition, this is just a bug and a fib.  It ignores what error happened and any diagnostic information.  It does not restore the old config if there was one. It sets init = false; so that it does not attempt to re-initialize.  The next thing that happens is a failure in distant code that assumes that the database is correctly connected.

File: extras/indexing/src/main/java/mvm/rya/indexing/FilterFunctionOptimizer.java (hunk 9 above)
Comment (DLotts) [inline extras/indexing/src/main/java/mvm/rya/indexing/FilterFunctionOptimizer.java]: Same as the first one.  Now redundant, and the catch is no longer required.

File: dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java (hunk 3 above)
Comment (DLotts) [inline dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java]: Private methods called from this method, correct?  Great idea.

**Comment by DLotts** [comment]
This is complete and ready to review.

**Comment by DLotts** [comment]
This was pulled into apache:develop !
